metricflow/plan_conversion/sql_join_builder.py

```diff
@@ -14,18 +15,14 @@
     SqlLogicalOperator,
     SqlTimeDeltaExpression,
 )
-from metricflow.time.time_granularity import TimeGranularity
 
 SqlDataSetT = TypeVar("SqlDataSetT", bound=SqlDataSet)
 
 
 def _make_cumulative_metric_join_condition(
-    metric_data_set_alias: str,
-    metric_data_set_metric_time_col: str,
-    time_spine_data_set_alias: str,
-    time_spine_data_set_time_col: str,
-    window: Optional[CumulativeMetricWindow],
-    grain_to_date: Optional[TimeGranularity],
+    metric_data_set: AnnotatedSqlDataSet,
+    time_spine_data_set: AnnotatedSqlDataSet,
+    node: JoinOverTimeRangeNode[SqlDataSetT],
```

Comment:
is there a reason we put a node here instead of using the window, grain_to_date arguments? Are we just restricting this to only be used by places that utilizes the JoinOverTimeRangeNode? I was wondering cause for conversion metrics, I did want to make use of this but it's not really joining to a time_spine, but rather joining against some time dimension between the 2 datasets. But other than that the entire windowing logic would be the same.

Reply:
I made the helper very explicitly about cumulative metrics because I found the JoinOverTimeRangeNode to really be a CumulativeMetricTimeSpineJoinNode - it's not readily generalizable because of the way it couples to the visitor method that converts it to SQL. We can - and probably should - refactor this along with the other 3-4 time window join use cases we're working on.

If you'd like to modify the name and signature of this helper and broaden its utility that's perfectly fine. Personally, I was thinking of getting everything in there and then figuring out how to make the design more sensible. We need a holistic approach to these time window joins (and likely to range joins more generally), because specifying which point in the range to favor is something users will need to be able to do. I could be very wrong about this, but the ad hoc "window" and "grain to date" properties in the cumulative metrics node don't feel like the right foundation to me.